File: uniffi_bindgen/src/bindings/kotlin/templates/CallbackInterfaceImpl.kt
{%- let trait_impl=format!("uniffiCallbackInterface{}", name) %}

// Put the implementation in an object so we don't pollute the top-level namespace
internal object {{ trait_impl }} {
    {%- for (ffi_callback, meth) in vtable_methods.iter() %}
    {%- if !meth.is_async() %}
    internal object {{ meth.name()|var_name }}: UniffiCallbackMethod {
        override fun callback(uniffiFfiBuffer: Pointer) {
            var uniffiArgCursor = UniffiBufferCursor(uniffiFfiBuffer)
            val uniffiMethodHandle = UniffiFfiSerializerHandle.read(uniffiArgCursor)
            {%- for arg in meth.arguments() %}
            val {{ arg.name()|var_name }} = {{ arg|ffi_serializer_name(ci) }}.read(uniffiArgCursor);
            {%- endfor %}
            val uniffiObj = {{ ffi_converter_name }}.handleMap.get(uniffiMethodHandle)

            try {
                {%- match meth.throws_type() %}
                {%- when None %}
                {% if meth.return_type().is_some() %}val uniffiReturnValue = {% endif %}uniffiObj.{{ meth.name()|fn_name() }}(
                    {%- for arg in meth.arguments() %}
                    {{ arg|lift_fn }}({{ arg.name()|var_name }}),
                    {%- endfor %}
                )
                {%- when Some(error_type) %}
                {% if meth.return_type().is_some() %}val uniffiReturnValue = {% endif %}try {
                    uniffiObj.{{ meth.name()|fn_name() }}(
                        {%- for arg in meth.arguments() %}
                        {{ arg|lift_fn }}({{ arg.name()|var_name }}),
                        {%- endfor %}
                    )
                } catch(e: {{ error_type|type_name(ci) }}) {
                    val uniffiReturnCursor = UniffiBufferCursor(uniffiFfiBuffer)
                    UniffiFfiSerializerUniffiRustCallStatus.write(
                        uniffiReturnCursor,
                        UniffiRustCallStatus.create(UNIFFI_CALL_ERROR, {{ error_type|lower_fn }}(e))
                    )
                    return
                }
                {%- endmatch %}

                val uniffiReturnCursor = UniffiBufferCursor(uniffiFfiBuffer)
                // Default RustCallStatus signals success
                UniffiFfiSerializerUniffiRustCallStatus.write(uniffiReturnCursor, UniffiRustCallStatus.ByValue())
                {%- if let Some(return_type) = meth.return_type() %}
                {{ return_type|ffi_serializer_name(ci) }}.write(uniffiReturnCursor, {{ return_type|lower_fn }}(uniffiReturnValue))
                {%- endif %}
            } catch(e: kotlin.Exception) {
                val uniffiReturnCursor = UniffiBufferCursor(uniffiFfiBuffer)
                try { 
                    val err = {{ Type::String.borrow()|lower_fn }}(e.stackTraceToString())
                    UniffiFfiSerializerUniffiRustCallStatus.write(
                        uniffiReturnCursor,
                        UniffiRustCallStatus.create(UNIFFI_CALL_UNEXPECTED_ERROR, err)
                    )
                } catch(_: Throwable) {
                    // Exception serializing the error message, just use an empty RustBuffer.
                    UniffiFfiSerializerUniffiRustCallStatus.write(
                        uniffiReturnCursor,
                        UniffiRustCallStatus.create(UNIFFI_CALL_UNEXPECTED_ERROR, RustBuffer.ByValue())
                    )
                }
            }
        }
    }
    {%- else %}
    internal object {{ meth.name()|var_name }}: UniffiAsyncCallbackMethod {
        override fun callback(uniffiFfiBuffer: Pointer, uniffiCallback: UniffiForeignFutureCallback): UniffiForeignFutureDroppedCallback {
            var uniffiArgCursor = UniffiBufferCursor(uniffiFfiBuffer)
            val uniffiMethodHandle = UniffiFfiSerializerHandle.read(uniffiArgCursor)
            {%- for arg in meth.arguments() %}
            val {{ arg.name()|var_name }} = {{ arg|ffi_serializer_name(ci) }}.read(uniffiArgCursor);
            {%- endfor %}
            val uniffiFutureHandle = UniffiFfiSerializerHandle.read(uniffiArgCursor)
            val uniffiObj = {{ ffi_converter_name }}.handleMap.get(uniffiMethodHandle)

            return uniffiCallAsync(uniffiFfiBuffer) {
                val returnFfiBuffer = Memory(
                    {% if let Some(return_ty) = meth.return_type() %}{{ return_ty|ffi_serializer_name(ci) }}.size() + {% endif -%}
                    UniffiFfiSerializerLong.size() + UniffiFfiSerializerUniffiRustCallStatus.size()
                )
                try {
                    {%- match meth.throws_type() %}
                    {%- when None %}
                    {%- if meth.return_type().is_some() %}val uniffiReturnValue = {% endif %}uniffiObj.{{ meth.name()|fn_name() }}(
                        {%- for arg in meth.arguments() %}
                        {{ arg|lift_fn }}({{ arg.name()|var_name }}),
                        {%- endfor %}
                    )
                    {%- when Some(error_type) %}
                    {% if meth.return_type().is_some() %}val uniffiReturnValue = {% endif %}try {
                        uniffiObj.{{ meth.name()|fn_name() }}(
                            {%- for arg in meth.arguments() %}
                            {{ arg|lift_fn }}({{ arg.name()|var_name }}),
                            {%- endfor %}
                        )
                    } catch(e: {{ error_type|type_name(ci) }}) {
                        val uniffiReturnCursor = UniffiBufferCursor(returnFfiBuffer)
                        UniffiFfiSerializerHandle.write(uniffiReturnCursor, uniffiFutureHandle)
                        UniffiFfiSerializerUniffiRustCallStatus.write(
                            uniffiReturnCursor,
                            UniffiRustCallStatus.create(UNIFFI_CALL_ERROR, {{ error_type|lower_fn }}(e))
                        )
                        uniffiCallback.callback(returnFfiBuffer)
                        return@uniffiCallAsync
                    }
                    {%- endmatch %}

                    val uniffiReturnCursor = UniffiBufferCursor(returnFfiBuffer)
                    UniffiFfiSerializerHandle.write(uniffiReturnCursor, uniffiFutureHandle)
                    // Default RustCallStatus signals success
                    UniffiFfiSerializerUniffiRustCallStatus.write(uniffiReturnCursor, UniffiRustCallStatus.ByValue())
                    {%- if let Some(return_type) = meth.return_type() %}
                    {{ return_type|ffi_serializer_name(ci) }}.write(uniffiReturnCursor, {{ return_type|lower_fn }}(uniffiReturnValue))
                    {%- endif %}
                    uniffiCallback.callback(returnFfiBuffer)
                } catch(e: kotlin.Exception) {
                    val uniffiReturnCursor = UniffiBufferCursor(returnFfiBuffer)
                    try { 
                        val err = {{ Type::String.borrow()|lower_fn }}(e.stackTraceToString())
                        UniffiFfiSerializerHandle.write(uniffiReturnCursor, uniffiFutureHandle)
                        UniffiFfiSerializerUniffiRustCallStatus.write(
                            uniffiReturnCursor,
                            UniffiRustCallStatus.create(UNIFFI_CALL_UNEXPECTED_ERROR, err)
                        )
                    } catch(_: Throwable) {
                        // Exception serializing the error message, just use an empty RustBuffer.
                        UniffiFfiSerializerHandle.write(uniffiReturnCursor, uniffiFutureHandle)
                        UniffiFfiSerializerUniffiRustCallStatus.write(
                            uniffiReturnCursor,
                            UniffiRustCallStatus.create(UNIFFI_CALL_UNEXPECTED_ERROR, RustBuffer.ByValue())
                        )
                    }
                    uniffiCallback.callback(returnFfiBuffer)
                }
            }
        }
    }
    {%- endif %}
    {%- endfor %}

    internal object uniffiFree: UniffiCallbackMethod {
        override fun callback(uniffiFfiBuffer: Pointer) {
            val argCursor = UniffiBufferCursor(uniffiFfiBuffer)
            val handle = UniffiFfiSerializerHandle.read(argCursor)
            {{ ffi_converter_name }}.handleMap.remove(handle)
        }
    }

    internal object uniffiClone: UniffiCallbackMethod {
        override fun callback(uniffiFfiBuffer: Pointer) {
            val argCursor = UniffiBufferCursor(uniffiFfiBuffer)
            val handle = UniffiFfiSerializerHandle.read(argCursor)
            val clonedHandle = {{ ffi_converter_name }}.handleMap.clone(handle)
            val returnCursor = UniffiBufferCursor(uniffiFfiBuffer)
            UniffiFfiSerializerHandle.write(returnCursor, clonedHandle)
        }
    }

    internal var vtable = {{ vtable|ffi_type_name(ci) }}(
        uniffiFree,
        uniffiClone,
        {%- for (ffi_callback, meth) in vtable_methods.iter() %}
        {{ meth.name()|var_name() }},
        {%- endfor %}
    )

    // Registers the foreign callback with the Rust side.
    // This method is generated for each callback interface.
    internal fun register(lib: UniffiLib) {
        vtable.write()
        lib.{{ ffi_init_callback.pointer_ffi_name() }}(vtable.pointer)
    }
}

File: uniffi_bindgen/src/bindings/ruby/templates/CallbackInterfaceImpl.rb
{#
// Template to generate a vtable implementation for a callback interface.
// This template is used by the `CallbackInterfaceTemplate.rb` (for pure callback interfaces)
// and `ObjectTemplate.rb` (for trait interfaces with ObjectImpl::CallbackTrait).
//
// Expected variable: cbi_name (the callback interface name string)
#}
{%- let cbi = ci.get_callback_interface_definition(cbi_name).unwrap() %}
{%- let vtable_def = cbi.vtable_definition() %}
{%- let vtable_methods = cbi.vtable_methods() %}
{%- let ffi_init_callback = cbi.ffi_init_callback() %}

# Callback interface vtable implementation for {{ cbi_name }}.
module UniffiCallbackInterface{{ cbi_name|class_name_rb }}

  {%- for (ffi_callback, method)  in vtable_methods.iter() %}

  {%- if method.is_async() %}
  # Async callback method for {{ method.name() }}
  {{ method.name()|enum_name_rb }}_CALLBACK = Proc.new do |uniffi_handle, {%- for arg in method.arguments() %} {{ arg.name()|var_name_rb }},{% endfor %} uniffi_future_callback, uniffi_callback_data, uniffi_out_dropped_callback|
    uniffi_obj = CallbackInterface{{ cbi_name|class_name_rb }}FfiConverter.lift uniffi_handle
    {%- call rb::make_call_proc(method) %}{% endcall %}
    {%- call rb::async_handle_success_proc(method) %}{% endcall %}
    {%- call rb::async_handle_error_proc(method) %}{% endcall %}
    {%- call rb::async_throws_dispatch(method) %}{% endcall %}
  end

  {%- else %}
  # Callback method for {{ method.name() }}
  {{ method.name()|enum_name_rb }}_CALLBACK = Proc.new do |uniffi_handle, {%- for arg in method.arguments() %} {{ arg.name()|var_name_rb }},{% endfor %} uniffi_out_return, uniffi_call_status|
    uniffi_obj = {{ self::canonical_name(cbi.as_type().borrow()) }}FfiConverter.lift uniffi_handle
    {%- call rb::make_call_proc(method) %}{% endcall %}
    {%- call rb::write_return_value_proc(method) %}{% endcall %}
    {%- call rb::sync_throws_dispatch(method) %}{% endcall %}
  end
  {%- endif %}
  {%- endfor %}

  # Free callback: removes the handle from the map.
  UNIFFI_FREE_CALLBACK = Proc.new do |uniffi_handle|
    {{ self::canonical_name(cbi.as_type().borrow()) }}FfiConverter.handle_map.remove uniffi_handle
  end

  # Clone callback: clones the handle in the map.
  UNIFFI_CLONE_CALLBACK = Proc.new do |uniffi_handle|
    {{ self::canonical_name(cbi.as_type().borrow()) }}FfiConverter.handle_map.clone_handle uniffi_handle
  end

  # Create the VTable struct instance.
  UNIFFI_VTABLE = UniFFILib::{{ vtable_def.name() }}.new
  UNIFFI_VTABLE[:uniffi_free] = UNIFFI_FREE_CALLBACK
  UNIFFI_VTABLE[:uniffi_clone] = UNIFFI_CLONE_CALLBACK
  {%- for (ffi_callback, method)  in vtable_methods.iter() %}
  UNIFFI_VTABLE[:{{ method.name() }}] = {{ method.name()|enum_name_rb }}_CALLBACK
  {%- endfor %}

  # Register the VTable with Rust.
  UniFFILib.{{ ffi_init_callback.name() }}(UNIFFI_VTABLE)
end
